### PR TITLE
Add some macros

### LINKED ISSUE #14: support labels macro to create labels easily
we may refer https://github.com/pingcap/tikv/blob/master/src/util/macros.rs#L66 directly and change a little.

### LINKED ISSUE #26: use macro to register global metics easily
Like 
```
register_counter!(HTTP_COUNTER, counter_opts);

// then we can use 
HTTP_COUNTER.inc();
```

### PR DESCRIPTION
**Pull request by overvenus**
Resolved #26 
Resolved #14 

Example:

```rust
lazy_static! {
    static ref COUNTER: Counter = register_counter_with!(
        opts!(
            "test_opt",
            "test_opt_help",
            labels!{
                "test"=> "hello",
                "foo"=> "bar",
                "ans"=> "42",
            }
        )
    );
}

fn main() {
    COUNTER.inc();
    println!("{:?}", COUNTER.get());
}
```

PTAL @siddontang @disksing @BusyJay

### CODE REVIEW

File: src/macros.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[macro_export]
+macro_rules! labels {
+    ($($KEY: expr => $VALUE: expr ,)+) => {
+        {
+            use std::collections::HashMap;
+
+            let mut lbs = HashMap::<String, String>::new();
```
Comment (siddontang) [inline src/macros.rs]: I prefer not fixing String type here, because in Vec metrics, we use &str.

File: src/macros.rs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    };
+
+    () => {
+        HashMap::<String, String>::new()
+    }
+}
+
+#[macro_export]
+macro_rules! opts {
+    ($NAME: expr, $HELP: expr, $LABELS: expr) => {
```
Comment (siddontang) [inline src/macros.rs]: should support without labels.

File: src/macros.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#[macro_export]
+macro_rules! labels {
```
Comment (siddontang) [inline src/macros.rs]: add test

File: src/macros.rs [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+#[macro_export]
+macro_rules! opts {
+    ($NAME: expr, $HELP: expr, $LABELS: expr) => {
+        $crate::Opts::new($NAME, $HELP).const_labels($LABELS)
+    }
+}
+
+#[macro_export]
```
Comment (siddontang) [inline src/macros.rs]: add test and comment. 

Must tell user that this function will panic if passing wrong options. or we can supply another must register macro instead.

File: src/macros.rs [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+macro_rules! register_counter_with {
```
Comment (siddontang) [inline src/macros.rs]: can we support like 
```
register_counter(name, help);
register_counter(name, help, labels);
register_counter(opts); 
```

File: src/macros.rs (hunk 1 above)
Comment (siddontang) [inline src/macros.rs]: If hard to do this, we can add const_label for String specially.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/macros.rs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+        let opts = histogram_opts!(name, help, labels!{"test" => "hello", "foo" => "bar",});
+        assert_eq!(opts.common_opts.const_labels.len(), 2);
+        assert!(opts.common_opts.const_labels.get("test").is_some());
+        assert_eq!(opts.common_opts.const_labels.get("test").unwrap(), "hello");
+
+        let opts = histogram_opts!(name, help, []);
+        assert_eq!(opts.buckets.len(), 0);
+
+        let opts = histogram_opts!(name, help, [Vec::from(&[1.0, 2.0] as &[f64])]);
```
Comment (siddontang) [inline src/macros.rs]: add a test with histogram line/exp bucket.

File: src/macros.rs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+    ( $ OPTS : expr ) => {
+        {
+            let counter = $crate::Counter::with_opts($OPTS).unwrap();
+            $crate::register(Box::new(counter.clone())).map(|_| counter)
+        }
+    }
+}
+
+#[macro_export]
+macro_rules! register_gauge_with {
```
Comment (siddontang) [inline src/macros.rs]: no need with now.

Comment (overvenus) [inline src/macros.rs]: rename to `register_gauge` ?

Comment (siddontang) [inline src/macros.rs]: I think ok.

File: examples/example_hyper.rs [hunk 8]
```diff
@@ -23,46 +24,30 @@ use hyper::server::{Server, Request, Response};
...
-
-        prometheus::register(Box::new(histogram.clone())).unwrap();
-
-        histogram
-    };
+    static ref HTTP_COUNTER: Counter = register_counter_with!(
+        opts!(
+            "example_http_requests_total",
+            "Total number of HTTP requests made.",
+            labels!{"handler"=> "all",}
```
Comment (siddontang) [inline examples/example_hyper.rs]: `"handler" =>`, leave a blank.

**Comment by siddontang** [comment]
LGTM
PTAL @BusyJay

File: src/macros.rs [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+
+            let mut lbs = HashMap::new();
+            $(
+                lbs.insert($KEY, $VALUE);
+            )+
+
+            lbs
+        }
+    }
+}
```
Comment (disksing) [inline src/macros.rs]: How about do it like this?
```
macro_rules! labels {
    ( $ ( $ KEY : expr => $ VALUE : expr ) , * ) => {
        {
            let mut lbs = HashMap::new();
            $(
                lbs.insert($KEY, $VALUE);
            )*
            lbs
        }
    }
}
```

Comment (overvenus) [inline src/macros.rs]: Consider `labels!{}`, then `mut` is unnecessary and clippy complains [`let_and_return`](https://github.com/Manishearth/rust-clippy/wiki#let_and_return)

**Comment by disksing** [comment]
LGTM

**Pull request merged**